build/sdk_update.go

```diff
@@ -0,0 +1,322 @@
+		return err
+	}
+	return f.Close()
+}
+
+// UpdateSDK updates SDK-managed files for the driver located at root.
+//
+// If DryRun option is set, the function would not update any files, and instead will
+// return ErrChangesRequired if there are any changes required.
+func UpdateSDK(root string, opt *UpdateOptions) error {
+	if opt == nil {
+		opt = &UpdateOptions{}
+	}
+
+	m, err := manifest.Load(filepath.Join(root, manifest.Filename))
+	if err != nil {
+		return err
+	}
+	c := &updater{
+		root: root, opt: *opt,
+		context: map[string]interface{}{
+			"Manifest": m,
+		},
+	}
+
+	for _, file := range skeleton.AssetNames() {
+		if err := c.processAsset(file); err != nil {
+			return err
+		}
+	}
+	d, err := NewDriver(root)
+	if err != nil {
+		return err
+	}
+	var changed bool
+	if c.opt.DryRun {
+		changed, err = d.ScriptChanged()
+	} else {
+		changed, err = d.Prepare()
+	}
+	if err != nil {
+		return err
+	} else if changed {
+		c.notifyChangedFile(ScriptName)
+	}
+
+	if c.opt.DryRun && c.changes > 0 {
+		return ErrChangesRequired
+	}
+
+	return nil
+}
+
+func (c *updater) processAsset(name string) error {
+	if name == manifestTpl {
+		// manifest is always managed by the driver developer
+		// the template is only for the driver init
+		return nil
+	}
+	overwrite := managedFiles[name] && !overwriteManagedFiles
+
+	if strings.HasSuffix(name, tplExt) {
+		return c.processTemplateAsset(name, c.context, overwrite)
+	}
+
+	return c.processFileAsset(name, overwrite)
+}
+
+func (c *updater) processFileAsset(name string, overwrite bool) error {
+	content := skeleton.MustAsset(name)
+	info := mustAssetInfo(name)
+
+	name = fixGitFolder(name)
+	return c.writeIfChanged(filepath.Join(c.root, name), content, info.Mode(), overwrite)
+}
+
+var funcs = map[string]interface{}{
+	"escape_shield":  escapeShield,
+	"toExportedName": toExportedName,
+}
+
+func (c *updater) processTemplateAsset(name string, v interface{}, overwrite bool) error {
+	tpl := string(skeleton.MustAsset(name))
+
+	t, err := template.New(name).Funcs(funcs).Parse(tpl)
+	if err != nil {
+		return err
+	}
+
+	name = fixGitFolder(name)
+	file := filepath.Join(c.root, name[:len(name)-len(tplExt)])
+
+	buf := bytes.NewBuffer(nil)
+	if err := t.Execute(buf, v); err != nil {
+		return err
+	}
+
+	info := mustAssetInfo(name)
+	return c.writeIfChanged(file, buf.Bytes(), info.Mode(), overwrite)
+}
+
+func (c *updater) writeIfChanged(file string, content []byte, m os.FileMode, overwrite bool) error {
+	f, err := os.Open(file)
+	if os.IsNotExist(err) {
+		c.notifyMissingFile(file)
+		return c.writeFile(file, content, m)
+	} else if err != nil {
+		return err
+	} else if !overwrite {
+		return nil
+	}
+
+	original, err := ioutil.ReadAll(f)
+	if err != nil {
+		return err
+	}
+
+	if bytes.Compare(original, content) == 0 {
+		return nil
+	}
+
+	c.notifyChangedFile(file)
+	return c.writeFile(file, content, m)
+}
+
+func (c *updater) writeFile(file string, content []byte, m os.FileMode) error {
```

Comment:
Actually, I realized there's another potential simplification option, which might also address your concerns about control flow:

```go
func (c *updater) writeFile(file string, content []byte, m os.FileMode) error {
	if c.opt.DryRun {
		return nil
	} else if err := os.MkdirAll(filepath.Dir(file), 0755); err != nil {
		return err
	}
	err := ioutil.WriteFile(file, content, m)
	if err != nil {
		return err
	}
	c.opt.Debug.printf("File %q has been written\n", file)
	rel, err := filepath.Rel(c.root, file)
	if err != nil {
		return err
	}
	if !strings.HasPrefix(rel, ".git"+string(filepath.Separator)) {
		git := exec.Command("git", "add", rel)
		git.Dir = c.root
		if out, err := git.CombinedOutput(); err != nil {
			c.opt.Warning.printf("cannot add a file to git: %v\n%s", err, string(out))
		}
	}
	return nil
}
```

What do you think?

Reply:
This looks better from both points of view :)